languages/sqlite/main.go

```diff
@@ -0,0 +1,99 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+
+	"github.com/kr/pty"
+	"golang.org/x/sys/unix"
+
+	"github.com/replit/prybar/utils"
+)
+
+var Instance = &SQLite{}
+
+type SQLite struct {}
+
+// constructConfigFile generates commands to configure the sqlite CLI.
+// It writes them to a temporary file and returns its pathname.
+func constructConfigFile(config *utils.Config) string {
+	f, err := ioutil.TempFile("", "sqlite-config")
+	if err != nil {
+		panic(err)
+	}
+	defer f.Close()
+
+	// main and continuation prompts
+	// TODO: this probably doesn't handle quotation marks properly
+	f.WriteString(fmt.Sprintf(".prompt '%s' '%s'\n", config.Ps1, config.Ps2))
+
+	return f.Name()
+}
+
+func disableEcho(file *os.File) {
+	fd := int(file.Fd())
+	termios, err := unix.IoctlGetTermios(fd, unix.TCGETS)
+	if err != nil {
+		panic(err)
+	}
+
+	termios.Lflag &^= unix.ECHO
+
+	if err := unix.IoctlSetTermios(fd, unix.TCSETS, termios); err != nil {
+		panic(err)
+	}
+}
+
+func Execute(config *utils.Config) {
+	configFile := constructConfigFile(config)
+	args := []string{"-init", configFile}
+	cmd := exec.Command("sqlite3", args...)
+
+	ptty, tty, err := pty.Open()
+	if err != nil {
+		panic(err)
+	}
+	disableEcho(ptty)
+
+	cmd.Stderr = tty
+	cmd.Stdin = tty
+	cmd.Stdout = tty
+
+	err = cmd.Start()
+	if err != nil {
+		panic(err)
+	}
+
+	// file to execute
+	if len(config.Args) > 1 {
+		panic("too many arguments")
+	}
+	if len(config.Args) == 1 {
+		fileToRun := config.Args[0]
+		ptty.WriteString(fmt.Sprintf(".read %s\n", fileToRun))
+	}
+	
+	// set up I/O
+	go io.Copy(os.Stderr, ptty)
+	if config.Quiet {
+		go io.Copy(os.Stdout, filter(ptty))
+	} else {
+		go io.Copy(os.Stdout, ptty)
+	}
+	io.Copy(ptty, os.Stdin)
+}
+
+// filter removes all output until we get a prompt
+func filter(src io.Reader) io.Reader {
```

Comment:
this function is pretty tricky, it'd be best to find some other way to silence the header.

This will mysteriously drop bytes if the scanner reads more than required (which will pretty much always happen). You'll have to return a reader that'll first return what's left in the scanner's buffer followed by actually reading from src.

This all seems a little too complex (and is not possible if we want to use the pty passed down from pid1), so i'd advise finding a way to make sqlite3 not output the header at all.

Reply:
Unfortunately I couldn't find a way to disable the output, but I'd certainly love a better way. Here's the SQLite CLI source (grep for "sqlite version %s"): https://sqlite.org/src/artifact/82f8a473c01662f5

I could patch it to add a flag to control that output, which should be pretty straightforward; it's just that we'd then have to stay up to date with upstream.

Reply:
Could we just pipe stderr through grep (which already implements the buffer handling correctly)?

Reply:
Going to try modifying sqlite's CLI to get us what we want